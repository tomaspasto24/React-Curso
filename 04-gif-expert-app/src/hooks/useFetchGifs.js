import { useEffect, useState } from "react";
import getGifs from "../helpers/getGifs";

const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // Si category cambia se ejecuta el callback del 1er argumento.
    useEffect( () => {
        getGifs(category)
            .then( imgs => {
                setState({
                    data: imgs,
                    loading: false
                })
            })
    }, [category]);

    return state;
}

export default useFetchGifs;