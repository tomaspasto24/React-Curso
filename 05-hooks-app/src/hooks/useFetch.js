import React, {useState, useEffect, useRef} from 'react'

export const useFetch = ( url ) => {

    const isMounted = useRef(true);
    
    useEffect( () => {

        // Cuando el componente se desmonte, se cambia isMounted por false. 
        return () => {
            isMounted.current = false;
        }

    }, [])

    const [state, setstate] = useState({
        loading: true,
        error: null,
        data: null
    });

    useEffect(() => {
        
        setstate({
            loading: true,
            error: null,
            data: null
        });

        fetch(url)
            .then( (resp) => resp.json() )
            .then( data => {

                if (isMounted.current)
                {
                    setstate({
                        loading: false,
                        error: null,
                        data,
                    });
                };

            })
            .catch( () => {
                setstate({
                    loading: false,
                    error: 'No se pudo cargar la informaciÃ³n',
                    data: null,
                });
            });

    }, [url])

    return state;
}
