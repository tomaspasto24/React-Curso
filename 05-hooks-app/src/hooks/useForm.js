import { useState } from "react/cjs/react.development"


export const useForm = ( initialState = {} ) => {

    const [formValues, setFormValues] = useState(initialState);

    const reset = () => {
        setFormValues( initialState );
    }

    // DesustructuraciÃ³n del target del evento que se ingresa.
    const handleInputChange = ( {target} ) => 
    {
        setFormValues({
            ...formValues,
            [target.name]: target.value
        });
    };

    return [ formValues, handleInputChange, reset];

};
