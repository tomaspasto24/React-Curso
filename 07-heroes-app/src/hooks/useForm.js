import React, { useState } from 'react';

export const useForm = ( initialState = {} ) => {
    
    const [value, setValue] = useState(initialState);

    const reset = () => {
        setValue( initialState );
    }

    // DesestructuraciÃ³n de objeto e.
    const handleInputChange = ({ target }) => {
        setValue({ 
            ...value,
            [target.name]: target.value
        })
    };

    return [value, reset, handleInputChange];
};
