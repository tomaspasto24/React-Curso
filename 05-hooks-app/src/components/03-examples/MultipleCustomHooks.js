import React from 'react'
import { useFetch } from '../../hooks/useFetch';
import '../02-useEffect/effects.css';
import { useCounter } from '../../hooks/useCounter';

export const MultipleCustomHooks = () => {

    const {counter, increment} = useCounter(1);

    const {loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);

    // !! son para que de retorne falso si es null.
    const { author, quote} = !!data && data[0];
    
    return (
        <div>
            <h1>Breaking Bad Quotes</h1>
            <hr />

            {
                loading
                ?
                (
                    <div className='alert alert-info text-center'>
                        Loading...
                    </div>
                )
                :
                (
                    <blockquote className='blockquote aling-right'>
                    <p className='mb-0'>{ quote }</p>
                    <footer className='blockquote-footer'>{ author }</footer>
                    </blockquote>                        
                )
            }

            { !loading && ( <button className='btn btn-primary' onClick={increment}>Siguiente quote</button> ) }
        </div>
    )
}
