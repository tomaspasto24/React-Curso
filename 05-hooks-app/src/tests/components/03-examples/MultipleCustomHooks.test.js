import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';

// mock simula que usa esa funci칩n, se debe especificcar con mockReturnValue() lo que va a devolver.
jest.mock('../../../hooks/useFetch');

describe('Pruebas sobre MultipleCustomHooks', () => {
    
    test('Debe mostrarse correctamente', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null,
        });

        const wrapper = shallow( <MultipleCustomHooks /> );

        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de mostrar la informaci칩n', () => {

        useFetch.mockReturnValue({
            data: [{
                author: 'Tom치s',
                quote: 'Prueba'
            }]
        }); 

        const wrapper = shallow( <MultipleCustomHooks /> );

        expect( wrapper.find('.alert').exists() ).toBe(false);
        expect( wrapper.find('.mb-0').text() ).toBe('Prueba');
        expect( wrapper.find('.blockquote-footer').text() ).toBe('Tom치s');
    });

});