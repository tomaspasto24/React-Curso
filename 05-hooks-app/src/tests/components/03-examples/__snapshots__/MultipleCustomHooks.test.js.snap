// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre MultipleCustomHooks Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</div>
`;
